src/test/chain_tests.cpp

```diff
@@ -41,6 +43,133 @@ const CBlockIndex* NaiveLastCommonAncestor(const CBlockIndex* a, const CBlockInd
+    // and asserts that its properties are suitable for performant
+    // ancestor search.
+    // It does so by imitating the `GetAncestor` function to
+    // look for an ancestor, and ensuring that an arbitrary
+    // ancestor is found in a number of steps that is much lower
+    // than simple linear traversal.
+
+    // Build a chain
+    const auto chain_size{200'000};
+    std::vector<std::unique_ptr<CBlockIndex>> block_index;
+    block_index.reserve(chain_size);
+    // Create genesis block
+    block_index.push_back(std::make_unique<CBlockIndex>());
+    for(auto i{1}; i < chain_size; ++i) {
+        auto new_index{std::make_unique<CBlockIndex>()};
+        new_index->pprev = block_index.back().get();
+        BOOST_CHECK(new_index->pprev);
+        new_index->nHeight = new_index->pprev->nHeight + 1;
+        new_index->BuildSkip();
+        block_index.push_back(std::move(new_index));
+    }
+
+    const auto runs{1000};
+    const auto min_distance{500};
+    FastRandomContext ctx(/*fDeterministic =*/true);
+    for (auto i{0}; i < runs; ++i) {
+        // pick a height in the second half of the chain
+        auto start_height = chain_size / 2 + ctx.randrange(chain_size / 2);
+        BOOST_REQUIRE(start_height > 0 && start_height < chain_size);
+        auto& start_p = block_index[start_height];
+        // pick a target height, earlier by at least min_distance
+        auto delta{min_distance + ctx.randrange(start_height - min_distance)};
+        auto target_height{start_height - delta};
+        BOOST_REQUIRE(target_height > 0 && target_height < chain_size && target_height < start_height);
+
+        // Perform traversal from start to target, skipping, as implemented in `GetAncestor`
+        const CBlockIndex* pindexWalk = start_p.get();
+        auto heightWalk = start_height;
+        auto iteration_count{0};
+        while (heightWalk > target_height) {
+            int heightSkip = GetSkipHeight(heightWalk);
+            int heightSkipPrev = GetSkipHeight(heightWalk - 1);
+            if (pindexWalk->pskip != nullptr &&
+                (heightSkip == target_height ||
+                (heightSkip > target_height && !(heightSkipPrev < heightSkip - 2 &&
+                                        heightSkipPrev >= target_height)))) {
+                // Only follow pskip if pprev->pskip isn't better than pskip->pprev.
+                pindexWalk = pindexWalk->pskip;
+                heightWalk = heightSkip;
+            } else {
+                assert(pindexWalk->pprev);
+                pindexWalk = pindexWalk->pprev;
+                heightWalk--;
+            }
+            ++iteration_count;
+        }
+
+        // Asserts on the number of iterations:
+        // always less than 100 and less than 5% of the distance
+        BOOST_CHECK_LT(iteration_count, 100);
+        BOOST_REQUIRE(delta > 0);
+        auto rel_iter_count = double(iteration_count) / double(delta);
+        BOOST_CHECK_LT(rel_iter_count, 0.05);
```

Comment:
can we avoid floating point arithmetic here?

```suggestion
        BOOST_CHECK_LT(iteration_count * 20, delta);
```